[dynamic shapes] prims_common non_overlapping_and_dense

pianpwk opened the pull request:
Differential Revision: D80120333




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

laithsakka commented on torch/_prims_common/__init__.py:
```diff
@@ -476,35 +465,78 @@ class K(NamedTuple):
-        def __ge__(self, other):
-            return guard_size_oblivious(self.stride >= other.stride)
+            return (
+                guard_or_false(self.stride < other.stride)  # checks statically known inequality
```
statically known-> backed and statically known

laithsakka commented on torch/_prims_common/__init__.py:
```diff
@@ -438,32 +438,21 @@ def is_channels_last_contiguous_or_false(a: Tensor) -> bool:
 
 
-def is_non_overlapping_and_dense(a: Tensor) -> bool:
+def _is_non_overlapping_and_dense(sizes, strides) -> bool:
```
call the function _is_non_overlapping_and_dense_or_false.
"""may return false instead of throwing DDE at data dependency"

laithsakka commented on test/test_dynamic_shapes.py:
```diff
@@ -896,6 +896,64 @@ def test_non_overlapping_and_dense_unbacked(self):
 
+    def test_prims_non_overlapping_and_dense(self):
+        shape_env = ShapeEnv()
+        cf = torch._prims_common.is_non_overlapping_and_dense
```
is_non_overlapping_and_dense-> is_non_overlapping_and_dense_or_false?

laithsakka commented on torch/_prims_common/__init__.py:
```diff
@@ -476,33 +489,49 @@ class K(NamedTuple):
-    for length, stride in lengths_and_strides:
-        if guard_or_false(length == 1):
-            continue
+def is_non_overlapping_and_dense(a: Tensor) -> bool:
```
is this for BC compatibility?
In my mind calling is_non_overlapping_and_dense should throw DDE
calling _is_non_overlapping_and_dense_or_false should not
now this is sort of alias for _is_non_overlapping_and_dense_or_false
meaning unbacked semantics are implicit to users calling is_non_overlapping_and_dense


maybe its ok

laithsakka commented on torch/_prims_common/__init__.py:
```diff
@@ -476,33 +489,49 @@ class K(NamedTuple):
+            # for backed symbols, this is practically a < operation
+            # for unbacked, we return True if < is statically known,
+            # then try to answer this symbolically, with stride ordering semantics
+            # (e.g. u0 < u0 is False, u0 < u1 is False with no axioms, u0 < 2 * u0 is True)
```
please add a comnent explaining why its ok if in the worst case for unbacked we picked a wrong order.

pianpwk committed:
[dynamic shapes] prims_common non_overlapping_and_dense (#160462)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/160462

Test Plan:
test_dynamic_shapes

Rollback Plan:

Reviewed By: laithsakka

Differential Revision: D80120333
